File: apps/web-admin/src/app/features/map/utils/map-fog-effect.ts
import type { AmbientTickOptions } from './map-ambient-zone';
import { parkPlanSize, parkParticleTarget } from './map-park-visual-scale';
import { ambientScreenScale } from './map-ambient-zone';
import { ambientWindVector, DEFAULT_AMBIENT_WIND } from './map-ambient-wind';

interface FogPatch {
  bx: number;
  by: number;
  vx: number;
  vy: number;
  r: number;
  phase: number;
}

export class MapFogEffect {
  private patches: FogPatch[] = [];
  private intensity = 0.35;
  private sizeMul = 1;
  private containsPoint: ((bx: number, by: number) => boolean) | null = null;

  setIntensity(value: number): void {
    this.intensity = Math.min(1, Math.max(0, value));
  }

  setSizeMul(value: number): void {
    this.sizeMul = Math.min(2.5, Math.max(0.08, value));
  }

  setContainsPoint(fn: ((bx: number, by: number) => boolean) | null): void {
    this.containsPoint = fn;
  }

  clear(): void {
    this.patches = [];
  }

  private inZone(bx: number, by: number): boolean {
    return !this.containsPoint || this.containsPoint(bx, by);
  }

  private spawn(bounds: AmbientTickOptions['bounds'], wind = DEFAULT_AMBIENT_WIND): FogPatch {
    const spanX = bounds.maxX - bounds.minX;
    const spanY = bounds.maxY - bounds.minY;
    let bx = bounds.minX + Math.random() * spanX;
    let by = bounds.minY + Math.random() * spanY;
    for (let t = 0; t < 20; t++) {
      bx = bounds.minX + Math.random() * spanX;
      by = bounds.minY + Math.random() * spanY;
      if (this.inZone(bx, by)) break;
    }
    const wv = ambientWindVector(wind);
    return {
      bx,
      by,
      vx: wv.vx * 0.35 + (Math.random() - 0.5) * 0.12,
      vy: wv.vy * 0.28 + (Math.random() - 0.5) * 0.1,
      r: parkPlanSize(18 + Math.random() * 32),
      phase: Math.random() * Math.PI * 2,
    };
  }

  tick(options: AmbientTickOptions, dt = 1): void {
    if (options.containsPoint) this.containsPoint = options.containsPoint;
    const wind = options.wind ?? DEFAULT_AMBIENT_WIND;
    const wv = ambientWindVector(wind);
    const bounds = options.bounds;
    const spanX = bounds.maxX - bounds.minX;
    const spanY = bounds.maxY - bounds.minY;
    if (spanX <= 0 || spanY <= 0) return;

    const inten = Math.max(0.15, this.intensity);
    const target = parkParticleTarget(3, 9, inten);

    while (this.patches.length < target) this.patches.push(this.spawn(bounds, wind));
    while (this.patches.length > target) this.patches.pop();

    for (const p of this.patches) {
      p.phase += 0.012 * dt;
      p.vx += (wv.vx * 0.35 - p.vx) * 0.03 * dt;
      p.vy += (wv.vy * 0.28 - p.vy) * 0.03 * dt;
      p.bx += p.vx * dt + Math.sin(p.phase) * 0.08 * dt;
      p.by += p.vy * dt + Math.cos(p.phase * 0.8) * 0.05 * dt;
      if (!this.inZone(p.bx, p.by) || p.bx < bounds.minX - 20 || p.bx > bounds.maxX + 20) {
        const n = this.spawn(bounds, wind);
        Object.assign(p, n);
      }
    }
  }

  draw(
    ctx: CanvasRenderingContext2D,
    clipPath: Path2D | null,
    toScreen: (bx: number, by: number) => { x: number; y: number },
    screenScale = 1,
  ): void {
    const inten = Math.max(0.2, this.intensity);
    const sr = ambientScreenScale(screenScale, this.sizeMul);
    ctx.save();
    if (clipPath) ctx.clip(clipPath);

    for (const p of this.patches) {
      if (!this.inZone(p.bx, p.by)) continue;
      const { x, y } = toScreen(p.bx, p.by);
      const pulse = 0.82 + Math.sin(p.phase) * 0.18;
      const radius = p.r * sr * pulse;
      // Nube puff plana (cartoon): varios discos cel con borde suave
      const puffs: Array<[number, number, number]> = [
        [-0.5, 0.12, 0.52],
        [0.5, 0.12, 0.52],
        [-0.9, 0.22, 0.36],
        [0.9, 0.22, 0.36],
        [0, -0.22, 0.62],
      ];
      const cel = (px: number, py: number, pr: number, a: number, col: string) => {
        const cx = x + px * radius;
        const cy = y + py * radius * 0.7;
        const cr = pr * radius;
        const g = ctx.createRadialGradient(cx, cy, cr * 0.7, cx, cy, cr);
        g.addColorStop(0, `rgba(${col}, ${a})`);
        g.addColorStop(0.8, `rgba(${col}, ${a})`);
        g.addColorStop(1, `rgba(${col}, 0)`);
        ctx.fillStyle = g;
        ctx.beginPath();
        ctx.arc(cx, cy, cr, 0, Math.PI * 2);
        ctx.fill();
      };
      for (const [px, py, pr] of puffs) cel(px, py, pr, inten * 0.16, '236, 244, 255');
      // realce superior
      cel(-0.2, -0.2, 0.42, inten * 0.12, '255, 255, 255');
    }

    ctx.restore();
  }
}

File: apps/mobile-android/app/src/main/java/com/univalle/pedrochacolla/ui/dashboard/MapFogEffect.kt
package com.univalle.pedrochacolla.ui.dashboard

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RadialGradient
import android.graphics.Shader
import com.univalle.pedrochacolla.utils.map.AmbientTickOptions
import com.univalle.pedrochacolla.utils.map.AmbientWind
import com.univalle.pedrochacolla.utils.map.MapAmbientZone
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

/** Niebla cartoon — port de `map-fog-effect.ts`. */
class MapFogEffect {
    var intensity = 0.35f
    var sizeMul = 1f
        set(value) { field = value.coerceIn(0.08f, 2.5f) }

    private data class Patch(
        var bx: Float, var by: Float, var vx: Float, var vy: Float,
        val r: Float, var phase: Float,
    )

    private val patches = mutableListOf<Patch>()
    private var containsPoint: ((Float, Float) -> Boolean)? = null
    private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    fun clear() { patches.clear() }
    fun setContainsPoint(fn: ((Float, Float) -> Boolean)?) { containsPoint = fn }

    private fun inZone(bx: Float, by: Float) = containsPoint?.invoke(bx, by) ?: true

    private fun spawn(bounds: AmbientTickOptions, wind: AmbientWind): Patch {
        val spanX = bounds.bounds.maxX - bounds.bounds.minX
        val spanY = bounds.bounds.maxY - bounds.bounds.minY
        var bx = bounds.bounds.minX + Random.nextFloat() * spanX
        var by = bounds.bounds.minY + Random.nextFloat() * spanY
        repeat(20) {
            bx = bounds.bounds.minX + Random.nextFloat() * spanX
            by = bounds.bounds.minY + Random.nextFloat() * spanY
            if (inZone(bx, by)) return@repeat
        }
        val (wx, wy) = MapAmbientZone.ambientWindVector(wind)
        return Patch(
            bx, by,
            wx * 0.35f + (Random.nextFloat() - 0.5f) * 0.12f,
            wy * 0.28f + (Random.nextFloat() - 0.5f) * 0.1f,
            MapAmbientZone.parkPlanSize(18f + Random.nextFloat() * 32f),
            Random.nextFloat() * 6.28f,
        )
    }

    fun tick(options: AmbientTickOptions, dt: Float = 1f) {
        options.containsPoint?.let { containsPoint = it }
        val wind = options.wind
        val bounds = options.bounds
        val spanX = bounds.maxX - bounds.minX
        val spanY = bounds.maxY - bounds.minY
        if (spanX <= 0f || spanY <= 0f) return
        val inten = maxOf(0.15f, intensity)
        val target = MapAmbientZone.parkParticleTarget(3, 9, inten)
        while (patches.size < target) patches.add(spawn(options, wind))
        while (patches.size > target) patches.removeAt(patches.lastIndex)
        val (wx, wy) = MapAmbientZone.ambientWindVector(wind)
        for (p in patches) {
            p.phase += 0.012f * dt
            p.vx += (wx * 0.35f - p.vx) * 0.03f * dt
            p.vy += (wy * 0.28f - p.vy) * 0.03f * dt
            p.bx += p.vx * dt + sin(p.phase) * 0.08f * dt
            p.by += p.vy * dt + cos(p.phase * 0.8f) * 0.05f * dt
            if (!inZone(p.bx, p.by) || p.bx < bounds.minX - 20f || p.bx > bounds.maxX + 20f) {
                val n = spawn(options, wind)
                p.bx = n.bx; p.by = n.by; p.vx = n.vx; p.vy = n.vy; p.phase = n.phase
            }
        }
    }

    fun draw(
        canvas: Canvas,
        clipPath: Path?,
        toScreen: (Float, Float) -> Pair<Float, Float>,
        screenScale: Float,
    ) {
        val inten = maxOf(0.2f, intensity)
        val sr = MapAmbientZone.ambientScreenScale(screenScale, sizeMul)
        canvas.save()
        clipPath?.let { canvas.clipPath(it) }
        val puffs = arrayOf(
            floatArrayOf(-0.5f, 0.12f, 0.52f),
            floatArrayOf(0.5f, 0.12f, 0.52f),
            floatArrayOf(0f, -0.22f, 0.62f),
        )
        for (p in patches) {
            if (!inZone(p.bx, p.by)) continue
            val (x, y) = toScreen(p.bx, p.by)
            val pulse = 0.82f + sin(p.phase) * 0.18f
            val radius = p.r * sr * pulse
            for (pr in puffs) {
                val cx = x + pr[0] * radius
                val cy = y + pr[1] * radius * 0.7f
                val cr = pr[2] * radius
                fillPaint.shader = RadialGradient(
                    cx, cy, cr,
                    intArrayOf(
                        android.graphics.Color.argb((inten * 41).toInt(), 236, 244, 255),
                        android.graphics.Color.argb(0, 236, 244, 255),
                    ),
                    floatArrayOf(0.7f, 1f),
                    Shader.TileMode.CLAMP,
                )
                canvas.drawCircle(cx, cy, cr, fillPaint)
            }
        }
        fillPaint.shader = null
        canvas.restore()
    }
}
